Clippy is broken with nightly

```shell
   Compiling clippy_lints v0.0.166
error[E0609]: no field `generics` on type `&rustc::hir::MethodSig`
  --> /Users/gnzlbg/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.166/src/lifetimes.rs:69:58
   |
69 |             check_fn_inner(cx, &sig.decl, Some(id), &sig.generics, item.span);
   |                                                          ^^^^^^^^

error[E0609]: no field `generics` on type `&rustc::hir::MethodSig`
  --> /Users/gnzlbg/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.166/src/lifetimes.rs:79:54
   |
79 |             check_fn_inner(cx, &sig.decl, body, &sig.generics, item.span);
   |                                                      ^^^^^^^^

error[E0609]: no field `generics` on type `&rustc::hir::MethodSig`
   --> /Users/gnzlbg/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.166/src/methods.rs:722:84
    |
722 |                    self_kind.matches(first_arg_ty, first_arg, self_ty, false, &sig.generics) {
    |                                                                                    ^^^^^^^^

error[E0609]: no field `generics` on type `&rustc::hir::MethodSig`
   --> /Users/gnzlbg/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.166/src/methods.rs:736:106
    |
736 |                     !self_kinds.iter().any(|k| k.matches(first_arg_ty, first_arg, self_ty, is_copy, &sig.generics)),
    |                                                                                                          ^^^^^^^^

error[E0609]: no field `generics` on type `&rustc::hir::MethodSig`
   --> /Users/gnzlbg/.cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.166/src/new_without_default.rs:111:21
    |
111 |             if !sig.generics.ty_params.is_empty() {
    |                     ^^^^^^^^

error: aborting due to 5 previous errors

error: Could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `clippy v0.0.166`, intermediate artifacts can be found at `/var/folders/mm/y444qqwd0kn52fycyz6nsph40000gn/T/cargo-install.nt8TXBcHowWR`

Caused by:
  build failed
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,10 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## 0.0.167
+* Rustup to *rustc 1.23.0-nightly (90ef3372e 2017-10-29)*
+* New lints: [`const_static_lifetime`], [`erasing_op`], [`fallible_impl_from`], [`println_empty_string`], [`useless_asref`]
+
 ## 0.0.166
 * Rustup to *rustc 1.22.0-nightly (b7960878b 2017-10-18)*
 * New lints: [`explicit_write`], [`identity_conversion`], [`implicit_hasher`], [`invalid_ref`], [`option_map_or_none`], [`range_minus_one`], [`range_plus_one`], [`transmute_int_to_bool`], [`transmute_int_to_char`], [`transmute_int_to_float`]
@@ -486,6 +490,7 @@ All notable changes to this project will be documented in this file.
 [`cmp_null`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#cmp_null
 [`cmp_owned`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#cmp_owned
 [`collapsible_if`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#collapsible_if
+[`const_static_lifetime`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#const_static_lifetime
 [`crosspointer_transmute`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#crosspointer_transmute
 [`cyclomatic_complexity`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#cyclomatic_complexity
 [`deprecated_semver`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#deprecated_semver
@@ -504,6 +509,7 @@ All notable changes to this project will be documented in this file.
 [`enum_glob_use`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#enum_glob_use
 [`enum_variant_names`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#enum_variant_names
 [`eq_op`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#eq_op
+[`erasing_op`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#erasing_op
 [`eval_order_dependence`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#eval_order_dependence
 [`expl_impl_clone_on_copy`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
 [`explicit_counter_loop`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#explicit_counter_loop
@@ -511,6 +517,7 @@ All notable changes to this project will be documented in this file.
 [`explicit_iter_loop`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#explicit_iter_loop
 [`explicit_write`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#explicit_write
 [`extend_from_slice`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#extend_from_slice
+[`fallible_impl_from`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#fallible_impl_from
 [`filter_map`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#filter_map
 [`filter_next`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#filter_next
 [`float_arithmetic`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#float_arithmetic
@@ -611,6 +618,7 @@ All notable changes to this project will be documented in this file.
 [`precedence`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#precedence
 [`print_stdout`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#print_stdout
 [`print_with_newline`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#print_with_newline
+[`println_empty_string`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#println_empty_string
 [`ptr_arg`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#ptr_arg
 [`pub_enum_variant_names`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#pub_enum_variant_names
 [`range_minus_one`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#range_minus_one
@@ -673,6 +681,7 @@ All notable changes to this project will be documented in this file.
 [`use_debug`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#use_debug
 [`use_self`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#use_self
 [`used_underscore_binding`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#used_underscore_binding
+[`useless_asref`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_asref
 [`useless_attribute`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_attribute
 [`useless_format`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_format
 [`useless_let_if_seq`]: https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_let_if_seq

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "clippy"
-version = "0.0.166"
+version = "0.0.167"
 authors = [
 	"Manish Goregaokar <manishsmail@gmail.com>",
 	"Andre Bogus <bogusandre@gmail.com>",
@@ -37,7 +37,7 @@ path = "src/driver.rs"
 
 [dependencies]
 # begin automatic update
-clippy_lints = { version = "0.0.166", path = "clippy_lints" }
+clippy_lints = { version = "0.0.167", path = "clippy_lints" }
 # end automatic update
 cargo_metadata = "0.2"
 

--- a/clippy_lints/Cargo.toml
+++ b/clippy_lints/Cargo.toml
@@ -1,7 +1,7 @@
 [package]
 name = "clippy_lints"
 # begin automatic update
-version = "0.0.166"
+version = "0.0.167"
 # end automatic update
 authors = [
 	"Manish Goregaokar <manishsmail@gmail.com>",

--- a/clippy_lints/src/lib.rs
+++ b/clippy_lints/src/lib.rs
@@ -93,12 +93,13 @@ pub mod entry;
 pub mod enum_clike;
 pub mod enum_glob_use;
 pub mod enum_variants;
-pub mod erasing_op;
 pub mod eq_op;
+pub mod erasing_op;
 pub mod escape;
 pub mod eta_reduction;
 pub mod eval_order_dependence;
 pub mod explicit_write;
+pub mod fallible_impl_from;
 pub mod format;
 pub mod formatting;
 pub mod functions;
@@ -106,7 +107,6 @@ pub mod identity_conversion;
 pub mod identity_op;
 pub mod if_let_redundant_pattern_matching;
 pub mod if_not_else;
-pub mod fallible_impl_from;
 pub mod infinite_iter;
 pub mod int_plus_one;
 pub mod invalid_ref;
@@ -209,7 +209,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
     let mut store = reg.sess.lint_store.borrow_mut();
     store.register_removed(
         "should_assert_eq",
-        "`assert!()` will be more flexible with RFC 2011"
+        "`assert!()` will be more flexible with RFC 2011",
     );
     store.register_removed(
         "extend_from_slice",
@@ -360,11 +360,11 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
 
     reg.register_lint_group("clippy_pedantic", vec![
         booleans::NONMINIMAL_BOOL,
-        const_static_lifetime::CONST_STATIC_LIFETIME,
         empty_enum::EMPTY_ENUM,
         enum_glob_use::ENUM_GLOB_USE,
         enum_variants::PUB_ENUM_VARIANT_NAMES,
         enum_variants::STUTTER,
+        fallible_impl_from::FALLIBLE_IMPL_FROM,
         if_not_else::IF_NOT_ELSE,
         infinite_iter::MAYBE_INFINITE_ITER,
         int_plus_one::INT_PLUS_ONE,
@@ -423,6 +423,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         booleans::LOGIC_BUG,
         bytecount::NAIVE_BYTECOUNT,
         collapsible_if::COLLAPSIBLE_IF,
+        const_static_lifetime::CONST_STATIC_LIFETIME,
         copies::IF_SAME_THEN_ELSE,
         copies::IFS_SAME_COND,
         copies::MATCH_SAME_ARMS,
@@ -441,6 +442,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         enum_variants::MODULE_INCEPTION,
         eq_op::EQ_OP,
         eq_op::OP_REF,
+        erasing_op::ERASING_OP,
         escape::BOXED_LOCAL,
         eta_reduction::REDUNDANT_CLOSURE,
         eval_order_dependence::DIVERGING_SUB_EXPRESSION,
@@ -455,7 +457,6 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         identity_conversion::IDENTITY_CONVERSION,
         identity_op::IDENTITY_OP,
         if_let_redundant_pattern_matching::IF_LET_REDUNDANT_PATTERN_MATCHING,
-        fallible_impl_from::FALLIBLE_IMPL_FROM,
         infinite_iter::INFINITE_ITER,
         invalid_ref::INVALID_REF,
         is_unit_expr::UNIT_EXPR,
@@ -509,6 +510,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         methods::SINGLE_CHAR_PATTERN,
         methods::STRING_EXTEND_CHARS,
         methods::TEMPORARY_CSTRING_AS_PTR,
+        methods::USELESS_ASREF,
         methods::WRONG_SELF_CONVENTION,
         minmax::MIN_MAX,
         misc::CMP_NAN,

--- a/clippy_lints/src/lifetimes.rs
+++ b/clippy_lints/src/lifetimes.rs
@@ -66,7 +66,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LifetimePass {
 
     fn check_impl_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx ImplItem) {
         if let ImplItemKind::Method(ref sig, id) = item.node {
-            check_fn_inner(cx, &sig.decl, Some(id), &sig.generics, item.span);
+            check_fn_inner(cx, &sig.decl, Some(id), &item.generics, item.span);
         }
     }
 
@@ -76,7 +76,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LifetimePass {
                 TraitMethod::Required(_) => None,
                 TraitMethod::Provided(id) => Some(id),
             };
-            check_fn_inner(cx, &sig.decl, body, &sig.generics, item.span);
+            check_fn_inner(cx, &sig.decl, body, &item.generics, item.span);
         }
     }
 }

--- a/clippy_lints/src/methods.rs
+++ b/clippy_lints/src/methods.rs
@@ -746,7 +746,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
                     if name == method_name &&
                        sig.decl.inputs.len() == n_args &&
                        out_type.matches(&sig.decl.output) &&
-                       self_kind.matches(first_arg_ty, first_arg, self_ty, false, &sig.generics) {
+                       self_kind.matches(first_arg_ty, first_arg, self_ty, false, &implitem.generics) {
                         span_lint(cx, SHOULD_IMPLEMENT_TRAIT, implitem.span, &format!(
                             "defining a method called `{}` on this type; consider implementing \
                              the `{}` trait or choosing a less ambiguous name", name, trait_name));
@@ -760,7 +760,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
                 for &(ref conv, self_kinds) in &CONVENTIONS {
                     if_chain! {
                         if conv.check(&name.as_str());
-                        if !self_kinds.iter().any(|k| k.matches(first_arg_ty, first_arg, self_ty, is_copy, &sig.generics));
+                        if !self_kinds.iter().any(|k| k.matches(first_arg_ty, first_arg, self_ty, is_copy, &implitem.generics));
                         then {
                             let lint = if item.vis == hir::Visibility::Public {
                                 WRONG_PUB_SELF_CONVENTION

--- a/clippy_lints/src/new_without_default.rs
+++ b/clippy_lints/src/new_without_default.rs
@@ -108,13 +108,12 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NewWithoutDefault {
                 // can't be implemented by default
                 return;
             }
-            //TODO: There is no sig.generics anymore and I don't know how to fix this.
-            //if !sig.generics.ty_params.is_empty() {
-            //    // when the result of `new()` depends on a type parameter we should not require
-            //    // an
-            //    // impl of `Default`
-            //    return;
-            //}
+            if !cx.generics.expect("method must have generics").ty_params.is_empty() {
+               // when the result of `new()` depends on a type parameter we should not require
+               // an
+               // impl of `Default`
+               return;
+            }
             if decl.inputs.is_empty() && name == "new" && cx.access_levels.is_reachable(id) {
                 let self_ty = cx.tcx
                     .type_of(cx.tcx.hir.local_def_id(cx.tcx.hir.get_parent(id)));

--- a/clippy_lints/src/new_without_default.rs
+++ b/clippy_lints/src/new_without_default.rs
@@ -108,12 +108,13 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NewWithoutDefault {
                 // can't be implemented by default
                 return;
             }
-            if !sig.generics.ty_params.is_empty() {
-                // when the result of `new()` depends on a type parameter we should not require
-                // an
-                // impl of `Default`
-                return;
-            }
+            //TODO: There is no sig.generics anymore and I don't know how to fix this.
+            //if !sig.generics.ty_params.is_empty() {
+            //    // when the result of `new()` depends on a type parameter we should not require
+            //    // an
+            //    // impl of `Default`
+            //    return;
+            //}
             if decl.inputs.is_empty() && name == "new" && cx.access_levels.is_reachable(id) {
                 let self_ty = cx.tcx
                     .type_of(cx.tcx.hir.local_def_id(cx.tcx.hir.get_parent(id)));

--- a/clippy_lints/src/utils/mod.rs
+++ b/clippy_lints/src/utils/mod.rs
@@ -924,6 +924,7 @@ pub fn opt_def_id(def: Def) -> Option<DefId> {
         Def::TyAlias(id) |
         Def::AssociatedTy(id) |
         Def::TyParam(id) |
+        Def::TyForeign(id) |
         Def::Struct(id) |
         Def::StructCtor(id, ..) |
         Def::Union(id) |

--- a/tests/ui/mut_mut.stderr
+++ b/tests/ui/mut_mut.stderr
@@ -40,12 +40,6 @@ error: generally you want to avoid `&mut &mut _` if possible
    |                 ^^^^^^^^^^^^^
 
 error: generally you want to avoid `&mut &mut _` if possible
-  --> $DIR/mut_mut.rs:30:17
-   |
-30 |         let y : &mut &mut u32 = &mut &mut 2;
-   |                 ^^^^^^^^^^^^^
-
-error: generally you want to avoid `&mut &mut _` if possible
   --> $DIR/mut_mut.rs:35:38
    |
 35 |         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
@@ -56,24 +50,6 @@ error: generally you want to avoid `&mut &mut _` if possible
    |
 35 |         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
    |                 ^^^^^^^^^^^^^^^^^^
-
-error: generally you want to avoid `&mut &mut _` if possible
-  --> $DIR/mut_mut.rs:35:22
-   |
-35 |         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
-   |                      ^^^^^^^^^^^^^
-
-error: generally you want to avoid `&mut &mut _` if possible
-  --> $DIR/mut_mut.rs:35:17
-   |
-35 |         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
-   |                 ^^^^^^^^^^^^^^^^^^
-
-error: generally you want to avoid `&mut &mut _` if possible
-  --> $DIR/mut_mut.rs:35:22
-   |
-35 |         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
-   |                      ^^^^^^^^^^^^^
 
 error: generally you want to avoid `&mut &mut _` if possible
   --> $DIR/mut_mut.rs:35:22

